src/Data/Text/Encoding.hs

```diff
@@ -213,45 +218,122 @@ isValidBS bs = start 0
+  | otherwise = runST $ do
+    marr <- A.new len'
+    outer marr len' 0 0
+    where
+      wordByteSize = sizeOf w
+
+      index :: Int -> Word8
+      index i
+        | i < len1  = B.index bs1 i
+        | otherwise = B.index bs2 $ i - len1
+
+      len :: Int
+      len = len1 + len2
+      len' :: Int
+      len' = (len `div` wordByteSize) + 4
+
+      -- outer :: (Bits w, Num w) => A.MArray s -> Int -> Int -> Int -> ST s (DecodeResult Text ByteString w)
+      outer dst dstLen = inner
+        where
+          inner srcOff dstOff
+            -- finished
+            | len - srcOff < 1 = goodSoFar
+            -- shortcut for utf-8
+            | otherwise =
+              case queryOptimization bs1 bs2 srcOff of
+                Just (arr, tLen) ->
+                  let minLen = tLen + dstOff in
+                  if minLen > dstLen
+                    then do
+                      let dstLen' = minLen + 4
+                      dst' <- A.resizeM dst dstLen'
+                      A.copyI tLen dst' dstOff arr 0
+                      outer dst' dstLen' (srcOff + tLen) minLen
+                    else do
+                      A.copyI tLen dst dstOff arr 0
+                      inner (srcOff + tLen) minLen
+                _ -> if len - srcOff < wordByteSize
+                    -- incomplete code point
+                    then goodSoFar
+                    else
+                      if dstOff + 4 > dstLen
+                        -- need more space in destination
+                        then do
+                          let dstLen' = dstLen + 4
+                          dst' <- A.resizeM dst dstLen'
+                          outer dst' dstLen' srcOff dstOff
+                        else
+                          case decodeF index len srcOff of
+                            WriteAndAdvance c advance -> do
+                              d <- unsafeWrite dst dstOff c
+                              inner (advance d) $ dstOff + d
+                            NeedMore -> goodSoFar
+                            Invalid -> invalid
+            where
+              contin off res = do
+                A.shrinkM dst dstOff
+                arr <- A.unsafeFreeze dst
+                pure . res (Text arr 0 dstOff) $ if off >= len1
+                  then B.drop (off - len1) bs2
+                  else B.drop off $ bs1 `B.append` bs2
+              goodSoFar =
+                contin srcOff $ \ t bs' ->
+                  DecodeResult t Nothing bs' srcOff
+              invalid =
+                let srcOff' = srcOff + wordByteSize
+                    bytesToWord n word =
+                      if n > 0
+                        then bytesToWord (n - 1) $ (fromIntegral . index $ srcOff + wordByteSize - n) .|. (word `shiftL` 8)
+                        else word
+                in
+                contin srcOff' $ \ t bs' ->
+                  DecodeResult t (Just $ bytesToWord wordByteSize 0) bs' srcOff'
```

Comment:
I thought about doing that and using a `Bool` instead of a `Maybe` to indicate a code point error. My hesitation is the type `DecodeResult !t !Bool !b !Int` implies that `Bool` could be `True` while `b` is empty. Admittedly, it can be ensured that such a thing is never the case, but it still feels wrong (for lack a better term).

Reply:
Isn't that already kind of an issue, in the case where the `Just w` is not actually the wrong word because it's actually the next one, one would have to go back and read from the bytestring to get the right diagnostic.

Reply:
You're right that it's not necessarily the problematic word of the code point, but it is the first word of the invalid code point (first non-decodable word?), even if the offending word follows two or three bytes later, and the documentation should reflect that. However, identifying the start of the code point was only one of purposes of `Just x`. The other purpose is that combined with rest of the unencoded `ByteString` it creates a non-empty `ByteString` type. Again, I wanted to avoid having the signature imply a possibly empty `ByteString` when there's an invalid code point. (I briefly considered using `DecodeResult Text (Either (w, ByteString) ByteString) Int` to emphasis this.)